python/isbn-verifier: add mentor notes (#2168)
exercism/website-copy: tracks/python/exercises/isbn-verifier/mentoring.md

# Mentoring

## Problem and Challenges

The problem asks you to write an ISBN validation checker.

* Any "-" in the input string should be ignored.
* A valid ISBN must be 10 values long.
* A valid ISBN must be entirely digits, other than the last character.
* A valid ISBN may have `"X"` as the last character, which should be treated as a `10` value.

> * The ISBN is valid if `sum(value * position_counter) % 11 == 0` for `value in <input string>` and `position_counter in <(10..1)>`.

Author: This was intended as pseudo-code. Or, as a compact and succient manner to encapsulate the requirements here. I could write it out in English in a long form.

> * The ISBN is valid if `sum(value * position_counter) % 11 == 0` for `value in <input string>` and `position_counter in <(10..1)>`.

Author: ```suggestion
* The ISBN is valid if the weighted sum of the digits is a multiple of 11 where the weight is the digit's value, ranging from 10 to 1.
```

> * The ISBN is valid if `sum(value * position_counter) % 11 == 0` for `value in <input string>` and `position_counter in <(10..1)>`.

Reviewer: I think this is better than the former version.  Otherwise use variable names, and then explain what those variable names reference.  Programmer are used to interpreting variable names (especially if they are well named) and inferring what is meant.

But I like the written out explanation.  I can see the value of having the code presented to show the applied description.  

So, both?

> * The ISBN is valid if `sum(value * position_counter) % 11 == 0` for `value in <input string>` and `position_counter in <(10..1)>`.

Author: If you want working code which demonstrates what is needed ... isn't that what the next section contains?

> * The ISBN is valid if `sum(value * position_counter) % 11 == 0` for `value in <input string>` and `position_counter in <(10..1)>`.

Reviewer: Then, yes, prefer the written here and the example in the next section.

> * The ISBN is valid if `sum(value * position_counter) % 11 == 0` for `value in <input string>` and `position_counter in <(10..1)>`.

Author: Done in this current iteration.

## Reasonable Solution

```python
def is_valid(isbn: str) -> bool:
    values = [value for value in isbn if value != "-"]

    if len(values) != 10:
        return False

    if values[-1] == "X":
        values[-1] = "10"

    if any(not value.isdigit() for value in values):
        return False

    total = sum(int(value) * (10 - position) for position, value in enumerate(values))
    return total % 11 == 0
```

## Talking Points

### Generator expressions

> The sum can be computed using a `for` loop or via the `sum()` [built-in function][builtins-functions].

Author: The `sum()` generator is in the "Reasonable Solution" section. Should it be repeated here? Referenced?

> The sum can be computed using a `for` loop or via the `sum()` [built-in function][builtins-functions].

Reviewer: You could cross reference it:  "See the Reasonable Solution section." for example.

> The sum can be computed using a `for` loop or via the `sum()` [built-in function][builtins-functions].

Author: ```suggestion
The sum can be computed using the `sum()` [built-in function][builtins-functions] (see "Reasonable Solution") or a `for` loop.
```

> The sum can be computed using a `for` loop or via the `sum()` [built-in function][builtins-functions].

Author: Done.

> The sum can be computed using a `for` loop or via the `sum()` [built-in function][builtins-functions].

Reviewer: ```suggestion
The sum can be computed using a `for` loop or via the `sum()` [built-in function][builtin-functions].
```

```python
total = 0
for position, value in enumerate(values)):
    total += int(value) * (10 - position)
```

Generator expressions are considered very Pythonic and enable compact and clean iteration.

> They are also very efficient.

Author: I could run benchmarks and dump values.
I could link to PyCon talks.
But this is guidance for mentors, like a research paper :)

Are you referring specifically to line 45? Or the entire argument in favor of using generators?

> They are also very efficient.

Reviewer: Specifically "They are also very efficient."  This could be "This is known to be more efficient than other potential approaches." Maybe.  But if there are references it is good to provide further reading to support this statement.  Otherwise we could pose it as a question for exploration.

> They are also very efficient.

Author: Couching it in more vague language shifts it from a "claim from authority" to an "appeal to the people". You're saying the same thing ... but saying "it's known" vs "I know". It doesn't change the substance in any way.

Generators store less state than iterators and require fewer lookups, which is why they're faster than iterators. Generators don't build a list (which requires list appends and storing all the elements) which is why they are faster than building a list. Should I include these details?

There is PEP 255 where generators are introduced; the PEP lays out the reason for generators which includes the inefficiencies of lists and iterators for this use case. Generators are introduced in part *because* they are more efficient. I suppose the PEP itself is a pretty good reference for their efficiency, if someone wanted to dig through it. (See Motivation, paragraphs 3 and 4.)

Or I could just state they are more efficient ;)\

> Otherwise we could pose it as a question for exploration

Are we posing the question to mentors? Suggesting mentors suggest students benchmark it?

References:

https://www.dabeaz.com/generators/ and his PyCon talks are an amazing watch/listen on generators. Googling for "python generators vs iterators" produces numerous results. https://www.python.org/dev/peps/pep-0255/ motivation.

See https://medium.com/analytics-vidhya/python-iterators-vs-iterables-vs-generators-e15b9403292e

> Generator is efficient for writing fast and compact code. This is an advantage over Python iterators. They are also simpler to code than do custom iterator.

> They are also very efficient.

Reviewer: So I think you are saying that you could reference PEP 255 where efficiencies are mentioned.  (More efficient in what?  CPU ticks, Memory, Power, Time complexity?) Or perhaps the references you show here are good items to show in the mentor notes?

> Are we posing the question to mentors?

This document can supply questions to pose to students, sometimes as copy/paste-able material.  And usually I avoid benchmark suggestions, unless the benchmarks are extremely simple, or if the student brings it up, and even them usually we find out if that is really the bottleneck in their solution, often by profiling instead, or at least before we benchmark.

> They are also very efficient.

Author: Reference to PEP-255 added.

### The position value

> The position value can be directly generated using `range()`.

Author: Adjusted.

> The position value can be directly generated using `range()`.

Author: ```suggestion
The position values can be directly generated using `range()`.
```

However, using `range()` requires mapping the position value to an ISBN positional value, by computing an index or via `zip()`.

```python
total = 0
for position, value in zip(range(10, 0, -1), values):
    total += int(value) * position
```

The [`enumerate()` function][builtins-functions] provides a nice way to get both a position and a value.
The position returned by `enumerate()` does need to be modified (`10 - position`) to be used.

### The trailing "X" value

The final value can be handled in the summation loop by checking the position and the value and replacing "X" with 10.
Doing so allows for keeping the ISBN as a `str` instead of converting it to `list[str]` to be able to hold a `"10"` value.

```python
total = 0
for position, value in enumerate(isbn):
    if position == 9 and value == "X":
        total += 10
    else:
        total += int(value) * (10 - position)
```

> Recognizing that changing the data from `str` to `list[str]` allows for more flexibility of what it can store (`"10"`) and a simpler summation loop.

Reviewer: ```suggestion
Recognize that changing the data from `str` to `list[str]` allows for more flexibility of what can be stored (`"10"`) and a simpler summation loop.
```

### Checking for non-digits

> The [built-in `any()` function][builtin-functions] is very convenient for any non-digit values in the input.

Reviewer: It is very convenient for any non-digit values in the input, but why?  How is it convenient?  What does it do for it?  It feels like there might be a word missing or perhaps a couple of words in this sentence.

Author: What would you like to see here? How about:
```suggestion
The [built-in `any()` function][builtin-functions] is a concise approach for checking for any non-digit values in the input.
```

Reviewer: That makes more sense to me.

If the input is first converted to a string and a trailing "X" is replaced with a digit, the digit check is fairly straight forward.
If the trailing "X" is not handled prior to the digit check, the final character needs to be handled separately.

```python
if any(not value.isdigit() for value in isbn[:-1]):
    return False

> if not isbn[-1].isdigit() and isbn[-1] != "X":

Reviewer: I tend to shy away from a double negative in the conditional statement, is there another way to express this with out the not and not equal?

Author: This isn't a double negative as the two negatives are applying to different conditionals. Do you prefer this?

```suggestion
if isbn[-1].isdigit() or isbn[-1] = "X":
    pass
else:
```

Reviewer: Yes, I prefer that, it is a more positive thing.  It also has the different body, from `return False` to just a `pass`.  Surprising to me, in a way, as I would have thought the other potential would have been `True`.  Because with the else, that should be the true branch, and I prefer true and then false conditional branches to return true and then false.  Otherwise I usually consider the conditional statement to be "backward".

Author: This is one of the many tests required to check if a string is a valid ISBN. All the tests need to pass before we can `return True`. I would argue that

```
if not condition:
    return False
```
is significantly more compact, readable and Pythonic than:
```
if condition:
    pass
else:
    return False
```

What if I wrapped it all in a single `not`?

```suggestion
if not (isbn[-1].isdigit() or isbn[-1] == "X"):
```

Reviewer: With the following line being `return False` I do question if that is correct, as the last "digit" is allowed to be "X".

It is part of the larger `any()` statement, so I will defer to you for the specifics here.  It is only that I wanted to consider options here.

Author: This line is separate from the `any()` statement. The last character of the ISBN is allowed to be `"X"`, per the problem spec.

    return False
```

Creating a list of values which can store `"10"` in place of `"X"` simplifies the digit check.

Another approach to checking for non-digits is to wrap the `int()` call in a `try-except`.
This allows leaving the input as a `str` if the last value is handled separately to support an `"X"` value.

```python
total = 0
for position, value in enumerate(isbn):
    if position == 9 and value == "X":
        total += 10
    else:
        try:
            total += int(value) * (10 - position)
        except ValueError:
	    return False
```

> Checking for non-digits separate from the summation allows for explicitly and directly handling that test.

Reviewer: ```suggestion
Checking for non-digits separately from the summation allows for explicitly and directly handling that test.
```

    total += int(value) * (10 - position)
```

Generator expressions are considered very Pythonic.

> Generators were added to the language (see [PEP-255]) becuase they provide a compact, clean and efficient method for iterating through data (compared to other approaches such as building lists or iterators).

Reviewer: ```suggestion
Generators were added to the language (see [PEP-255]) because they provide a compact, clean and efficient method for iterating through data (compared to other approaches such as building lists or iterators).
```
